src/protocol/prf_sharding/mod.rs

```diff
@@ -408,32 +475,82 @@ where
         // This is incorrect in the case that the CAP is less than the maximum value of "trigger value" for a single row
         // Not a problem if you assume that's an invalid input
         difference_to_cap: BitDecomposed::new(vec![Replicated::ZERO; TV::BITS as usize]),
+        source_event_timestamp: BitDecomposed::new(vec![Replicated::ZERO; TS::BITS as usize]),
     }
 }
 
 ///
 /// To support "Last Touch Attribution" we move the `breakdown_key` of the most recent source event
 /// down to all of trigger events that follow it.
 ///
-/// The logic here is extremely simple. For each row:
-/// (a) if it is a source event, take the breakdown key bits.
-/// (b) if it is a trigger event, take the breakdown key bits from the preceding line
-///
 async fn breakdown_key_of_most_recent_source_event<C>(
     ctx: C,
     record_id: RecordId,
     is_trigger_bit: &Replicated<Gf2>,
     prev_row_breakdown_key_bits: &BitDecomposed<Replicated<Gf2>>,
     cur_row_breakdown_key_bits: &BitDecomposed<Replicated<Gf2>>,
 ) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    field_of_most_recent_source_event(
+        ctx,
+        record_id,
+        is_trigger_bit,
+        prev_row_breakdown_key_bits,
+        cur_row_breakdown_key_bits,
+    )
+    .await
+}
+
+/// Same as above but for timestamps. If `attribution_window_seconds` is `None`, just
+/// return the previous row's timestamp. The bits aren't used but saves some multiplications.
+async fn timestamp_of_most_recent_source_event<C>(
+    ctx: C,
+    record_id: RecordId,
+    attribution_window_seconds: Option<NonZeroU32>,
+    is_trigger_bit: &Replicated<Gf2>,
+    prev_row_timestamp_bits: &BitDecomposed<Replicated<Gf2>>,
+    cur_row_timestamp_bits: &BitDecomposed<Replicated<Gf2>>,
+) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    match attribution_window_seconds {
+        None => Ok(prev_row_timestamp_bits.clone()),
+        Some(_) => {
+            field_of_most_recent_source_event(
+                ctx,
+                record_id,
+                is_trigger_bit,
+                prev_row_timestamp_bits,
+                cur_row_timestamp_bits,
+            )
+            .await
+        }
+    }
+}
+
+/// Move a field of the most recent source event down to all of trigger events that follow it.
+///
+/// The logic here is extremely simple. For each row:
+/// (a) if it is a source event, take the current field.
+/// (b) if it is a trigger event, take the field from the preceding line
+async fn field_of_most_recent_source_event<C>(
```

Comment:
Instead of "Field" perhaps call it "Attribute" since the former is kind of an overloaded term.